sgkit/io/vcf/vcf_writer.py

```diff
@@ -0,0 +1,291 @@
+
+    n_variants = ds.dims["variants"]
+    n_samples = ds.dims["samples"]
+
+    with open(output, mode="w") as out:
+        print(header_str, end="", file=out)
+
+        header_info_fields = _info_fields(header_str)
+        header_format_fields = _format_fields(header_str)
+
+        # fixed fields
+
+        contigs = np.array(ds.attrs["contigs"], dtype="S")
+        filters = np.array(ds.attrs["filters"], dtype="S")
+
+        chrom = ds.variant_contig.values
+        pos = ds.variant_position.values
+        id = ds.variant_id.values.astype("S")
+        alleles = ds.variant_allele.values.astype("S")
+        qual = ds.variant_quality.values
+        filter_ = ds.variant_filter.values
+
+        # info fields
+
+        # preconvert all info fields to byte representations
+        info_bufs = []
+        info_mask = np.full((len(header_info_fields), n_variants), False, dtype=bool)
+        info_indexes = np.zeros(
+            (len(header_info_fields), n_variants + 1), dtype=np.int32
+        )
+
+        k = 0
+        info_prefixes = []  # field names followed by '=' (except for flag/bool types)
+        for key in header_info_fields:
+            if f"variant_{key}" not in ds:
+                continue
+            if ds[f"variant_{key}"].dtype in (
+                np.int8,
+                np.int16,
+                np.int32,
+                np.float32,
+            ) or ds[f"variant_{key}"].dtype.kind in ("S", "O"):
+                values = ds[f"variant_{key}"].values
+                if values.dtype.kind == "O":
+                    values = values.astype("S")  # convert to fixed-length strings
+                info_mask[k] = create_mask(values)
+                info_bufs.append(
+                    np.empty(vcf_values_to_byte_buf_size(values), dtype=np.uint8)
+                )
+                vcf_values_to_byte_buf(info_bufs[k], 0, values, info_indexes[k])
+                info_prefixes.append(key + "=")
+                k += 1
+            elif ds[f"variant_{key}"].dtype == np.bool:
+                values = ds[f"variant_{key}"].values
+                info_mask[k] = create_mask(values)
+                info_bufs.append(np.zeros(0, dtype=np.uint8))
+                # info_indexes contains zeros so nothing is written for flag/bool
+                info_prefixes.append(key)
+                k += 1
+
+        info_mask = info_mask[:k]
+        info_indexes = info_indexes[:k]
+
+        info_prefixes = np.array(info_prefixes, dtype="S")
+
+        # format fields
+
+        def fkey_to_var(key):
+            return "call_genotype" if key == "GT" else f"call_{key}"
+
+        # these can have different sizes for different fields, so store in sequences
+        format_values = []
+        format_bufs = []
+
+        format_mask = np.full(
+            (len(header_format_fields), n_variants), False, dtype=bool
+        )
+
+        k = 0
+        format_fields = []
+        has_gt = False
+        for key in header_format_fields:
```

Comment:
Ditto here, could just assign key to variable rather than calling function several times.